### DISCUSSION TITLE
Tiff support

### DISCUSSION THREAD

**Discussion by alexbejann** [ngageoint/projections-ios#8] [General]
Hi @bosborn, 
Is it possible to add libtiff as a dependency to the PROJ so that pipelines with grids can be processed?

**Comment by bosborn** [comment]
That might be possible.  Would have to see if [building the PROJ source code](https://proj.org/en/9.2/install.html#compilation-and-installation-from-source-code) with [libtiff](https://proj.org/en/9.2/install.html#cmdoption-arg-TIFF_LIBRARY_RELEASE) was still [buildable and releasable with CocoaPods](https://github.com/ngageoint/PROJ/wiki/CocoaPods).

I don't know a ton about PROJ w/ libtiff if you don't choose to disable it during build time.  If grids are not available, does it default to the base behavior?  We would have to make sure it still functions w/o access to local or remote grids for use on both online and offline mobile applications.

**Reply by alexbejann** [reply]
That's something that I was trying to do myself. I tried to reproduce what you have done for PROJ by compiling and adding the ".podspec" and then adding enabling TIFF support in PROJ. This last part I am not successful. 

It does default to base behavior. The only thing it does it adds the ".tif" (grids) support. If you provide a pipeline string without grids it's fine. If you provide a pipeline string with grids you have to have them somewhere. Usually that's done with this [method](https://proj.org/en/9.2/development/reference/functions.html#c.proj_context_set_search_paths). 

I would like to help in adding this feature/support. 

Thanks for taking time answering!

**Reply by bosborn** [reply]
I wrote an internal ticket but I'm not sure when I'll get time next to mess with it.  The [wiki page I linked](https://github.com/ngageoint/PROJ/wiki/CocoaPods) are my notes to get to where it is at now.  The changes I had to make can be seen [here](https://github.com/ngageoint/PROJ/compare/08f394e...9.3.0).  If you want to play around with it, you could fork and work [off this branch](https://github.com/ngageoint/PROJ/tree/9.3.0_release) with the added libtiff built in.  You could try to `pod lib lint` at least.  Or I guess you could try to set up a [9.4.0 version](https://github.com/OSGeo/PROJ/releases/tag/9.4.0) branch and submit a pull request.  Otherwise I'll try to get to it when I can.

**Reply by alexbejann** [reply]
Sure, i will give it a try. Also, for the libtiff shouldn't be added a podspec file? Just like you did for PROJ-include

**Reply by bosborn** [reply]
Yeah that is possible.  I wasn't sure if building with libtiff included the compiled source it needed or if it still expects that dependency.  But also libtiff is not in CocoaPods to pull from meaning it might require a manual library link by a utilizing application (like through Xcode binary linking).  The downside being it would require anyone wanting to use PROJ w/ libtiff to configure it themselves.  Not sure how easy it would be to publish [libtiff](https://github.com/libsdl-org/libtiff) as a pod like we are doing with PROJ, but that would make it the most usable in the end.

**Reply by alexbejann** [reply]
Hey @bosborn, I think the best approach would be to release libtiff on Cocoapods. After to release another version of PROJ which has the libtiff enabled. I would propose to name it PROJ-tiff (or similar) because I changed the source file a bit. What do you think? I will add the link to the repos to this thread.

**Reply by bosborn** [reply]
That would probably be the most useful solution.  If it is just LibTIFF with the source edited to work with CocoaPods, could just name it LibTIFF.  Enable LibTIFF usage directly through CocoaPods.  Unless those source changes are PROJ specific somehow.

**Reply by alexbejann** [reply]
I read your comment moments after I pushed libtiff to Cocoapods. 

What I did is the following. 
I released [libtiff](https://github.com/alexbejann/libtiff/tree/4.6.2_release)
I created a new fork from your [PROJ](https://github.com/ngageoint/PROJ) fork and adjusted the files a bit by adding the libtiff dependency. So we have now [PROJ-LibTIFF](https://github.com/alexbejann/PROJ/tree/9.3.1_release) on cocoapods.

I had to make some adjustments for PROJ to get it working with libtiff. Also, the library itself complained about a lot of things. Should we include this PROJ-LibTIFF under the NGAGEOPOINT? 

Concluding
In this way I was able to achieve what I wanted in iOS. 
The libraries server as a base for a Flutter Plugin for an application that I am working on.

**Reply by bosborn** [reply]
Nice work.  So with the libtiff, there were more complaints than normal?  Console warnings and such?
Are you thinking we'd need both a PROJ release and a PROJ-LibTIFF going forward where someone can choose one or the other?  Or will a single PROJ release w/ libtiff added suffice?
Are you wanting to maintain the libtiff release going forward or hoping to get it pulled into a ngageoint repo?

**Reply by alexbejann** [reply]
Thanks! 
There were indeed more and I had to remove/comment out some files that were not used. I didn't include the [port folder](https://github.com/alexbejann/libtiff/tree/4.6.2_release/port) from libtiff. For PROJ I manually enforced libtiff in the [grids.cpp](https://github.com/alexbejann/PROJ/blob/9.3.1_release/src/grids.cpp) and everywhere **_TIFF_ENABLED_** was used. 

At the moment I don't see a way to _**DYNAMICALLY**_ enable libtiff in PROJ. Unless you have an idea how to do that and I will implement it, maybe we can create a pr in the official repo. Because the proj_config.h would contain that `#define TIFF_ENABLED 1` but that is also set from [CMAKE file](https://github.com/alexbejann/PROJ/blob/a1dcc48a92320bfc872282f75b25108b7f9e177d/CMakeLists.txt#L196C1-L197C1)

The proposal to put PROJ-LibTIFF and libtiff under ngageopoint repo is to give confidence to anyone wanting to use it :D, even if it's under your organization, if you could allow me to maintain them I would be more than happy. Any thoughts on that? I am open to keep them under my github profile as well.

**Reply by alexbejann** [reply]
I took and adjusted the [wiki page](https://github.com/ngageoint/PROJ/wiki/CocoaPods) that you made to show what I did for [PROJ-LibtTIFF](https://github.com/alexbejann/PROJ/wiki/CocoaPods) and [libtiff](https://github.com/alexbejann/libtiff)

**Reply by bosborn** [reply]
I guess keeping them separate pods seems best for now.  I can probably at least pull some of those changes in to have.  If you want to maintain ownership of the pod and publish updates that is a good option for now.  Shared ownership would probably be a harder sell with NGA and contracts.  Pod ownership is not great.  First to publish becomes the owner and then can add other contributors.  Vs the Maven Central Repo where an org can have contributors and publish a library under their package path (vs other org versions).  Publishing a ngageoint version, if we think that is the best way to go forward, is probably easiest by changing the pod name slightly.

**Reply by alexbejann** [reply]
Sure, I will maintain it. Yes, please feel free to pull/change. Hopefully, this will be improved in the future. If I can contribute to anything let me know.
Also, do you have any steps to get the library to work with swift projects? When importing the custom library created (in objective C) in the swift project is complaining about `PROJ-LibTIFF/proj.h` is not available. In your repo, I see you have created `.swift` file only with the Foundation import.

**Reply by bosborn** [reply]
The empty swift files were for defining iOS modules.  Using the libraries from both Objective-C and Swift apps with minimal configurations required it.  If a Podfile contains `use_modular_headers!`, modular_headers for the PROJ type library need to turned off to build: `pod 'PROJ', :modular_headers => false`.

I will pull some of your changes in (build file, wiki, etc) when I get back to working on iOS.

**Reply by alexbejann** [reply]
Cool! What if you ```use_modular_headers!``` and ```use_frameworks!```? Asking as for another project where I imported the ```proj-ios``` I was getting the ```Use of undeclared identifier 'nullptr'``` for some files. Same as the previous discussion that i opened in this repo.

**Reply by bosborn** [reply]
We dropped `use_frameworks!` from our Podfiles in both Objective-C and Swift projects.  CocoaPods [added static library support](https://blog.cocoapods.org/CocoaPods-1.5.0/).  Can you do that?

**Reply by alexbejann** [reply]
Yep, did that. But after removing the ```use_frameworks!``` and ```use_modular_headers!``` everything hangs at the built for some reason. I don't know of any way of allowing specific pods to be used with/as ```frameworks```

**Reply by alexbejann** [reply]
This turned out to be a misconfiguration of the iOS project. 

By using in my case

`pod 'PROJ-LibTIFF', :modular_headers => false`
`pod 'libtiff', :modular_headers => false`

and adjusting the Flutter project to remove `use_frameworks!` 

Relevant comment that helped solving the issue: 
https://github.com/flutter/flutter/issues/25676#issuecomment-532464440

**Comment by alexbejann** [comment]
The solution for this can be found in the fork that I created. Below are links to the Wikis of both repos

[PROJ-LibTiff](https://github.com/alexbejann/PROJ/wiki/CocoaPods#proj-libtiff)
[libtiff](https://github.com/alexbejann/libtiff/wiki/CocoaPods)
